microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineThroughputBenchmark.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.microbench.handler.ssl;
+
+import io.netty.util.internal.ThreadLocalRandom;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.TearDown;
+
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLException;
+import java.nio.ByteBuffer;
+
+public abstract class AbstractSslEngineThroughputBenchmark extends AbstractSslEngineBenchmark {
+
+    @Param({ "64", "128", "512", "1024", "4096" })
+    public int messageSize;
+
+    protected ByteBuffer wrapSrcBuffer;
+    protected ByteBuffer wrapDstBuffer;
+
+    @Setup(Level.Iteration)
+    public final void setup() throws Exception {
+        initEngines();
+        initHandshakeBuffers();
+
+        wrapDstBuffer = allocateBuffer(clientEngine.getSession().getPacketBufferSize());
+        wrapSrcBuffer = allocateBuffer(messageSize);
+
+        byte[] bytes = new byte[messageSize];
+        ThreadLocalRandom.current().nextBytes(bytes);
+        wrapSrcBuffer.put(bytes);
+        wrapSrcBuffer.flip();
+
+        // Complete the initial TLS handshake.
+        doHandshake();
+
+        doSetup();
+    }
+
+    protected void doSetup() throws Exception { }
+
+    @TearDown(Level.Iteration)
+    public final void tearDown() throws Exception {
+        destroyEngines();
+        destroyHandshakeBuffers();
+        freeBuffer(wrapSrcBuffer);
+        freeBuffer(wrapDstBuffer);
+        doTearDown();
+    }
+
+    protected void doTearDown() throws Exception { }
+
+    @Setup(Level.Invocation)
```

Comment:
Agreed that it is ideal from a logical perspective to have it completely isolated from the benchmark.  However I'm curious if we fall into the warnings that are listed in the javadocs for `Level.Invocation`.

As long as the overhead is consistent and "relatively small" compared to what we want to run in the benchmark then conceivably this may make the overhead of managing the counters/timers lower, reduce risk for coordinated omission, etc... WDYT?